apptest/tests/vmauth_routing_test.go

```
@@ -0,0 +1,77 @@
+package tests
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/apptest"
+)
+
+func TestVMAuthRouting(t *testing.T) {
+	tc := apptest.NewTestCase(t)
+	defer tc.Stop()
+
+	var proxiedRequestsCount int
+	backend := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		proxiedRequestsCount++
+	}))
+	defer backend.Close()
+
+	authConfig := fmt.Sprintf(`
+unauthorized_user:
+   url_prefix: %s
+  `, backend.URL)
+
+	const (
+		// it's not possible to use random ports
+		// it makes test flaky
+		listenPortPublic  = "50127"
+		listenPortPrivate = "50126"
+	)
+
+	vmauthFlags := []string{
+		fmt.Sprintf("-httpListenAddr=127.0.0.1:%s", listenPortPublic),
+		fmt.Sprintf("-httpInternalListenAddr=127.0.0.1:%s", listenPortPrivate),
+		"-flagsAuthKey=protected",
+	}
+	vmauth := tc.MustStartVmauth("vmauth",
+		vmauthFlags,
+		authConfig)
+
+	var hc http.Client
+	makeGetRequestExpectCode := func(targetURL string, expectCode int) {
+		t.Helper()
+		req, err := http.NewRequest("GET", targetURL, nil)
+		if err != nil {
+			t.Fatalf("cannot build http.Request for target=%q: %s", targetURL, err)
+		}
+		resp, err := hc.Do(req)
+		if err != nil {
+			t.Fatalf("unexpected http request error: %s", err)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != expectCode {
+			responseText, err := io.ReadAll(resp.Body)
```

Let's read the rentire response body unconditionally before comparing the status code I this allows to reuse the connection. Also fail fast with Fatalf() since this is an unexpected error that prevents the test from proceeding.